### ISSUE TITLE
Node.js support

### ISSUE THREAD

**Issue by rettgerst** [cytoscape/cytoscape.js#3230]
**Description of new feature**

_What should the new feature do?  For visual features, include an image/mockup of the expected output._

library should support node.js (by allowing the user to pass a jsdom element as a container)

**Motivation for new feature**

_Describe your use case for this new feature._

we want to send out a graph render in emails, but using node.js to render the graph rather than a headless browser.

**For reviewers**

_Reviewers should ensure that the following tasks are carried out for incorporated issues:_

- [ ] Ensure that the reporter has adequately described their idea.  If not, elicit more information about the use case.  You should iteratively build a spec together.
- [ ] Ensure that the issue is a good fit for the core library.  Some things are best done in extensions (e.g. UI-related features that aren't style-related).  Some things are best done by app authors themselves -- instead of in Cytoscape libraries.
- [ ] The issue has been associated with a corresponding [milestone](https://github.com/cytoscape/cytoscape.js/milestones).
- [ ] The commits have been incorporated into the `unstable` branch via pull request.  The corresponding pull request is cross-referenced.

**Comment by mikekucera** [comment]
Can you please describe how this would work in more detail. Currently cytoscape.js can be run under node in a headless mode, and it can render graphs on the server using cytosnap. Can you describe how your proposal differs, and why cytosnap does not meet your requirements?

There is currently some preliminary work being done on exporting graphs to PDF, which should also work in headless mode. Would that satisfy your requirements?

**Comment by PilsManuel** [comment]
I was researching node.js support for cytoscape when I stumbled upon this issue.

Unfortunately cytosnap is not an option as it seems to be not maintained anymore (last release 5 years ago) and has 4 high/critical security alerts according to [socket.dev](https://socket.dev/npm/package/cytosnap).

**Comment by rettgerst** [comment]
> Can you please describe how this would work in more detail. Currently cytoscape.js can be run under node in a headless mode, and it can render graphs on the server using cytosnap. Can you describe how your proposal differs, and why cytosnap does not meet your requirements?

my main interest in this change is the ability to reuse existing cytoscape-consuming code in node,js, without the added complexity, latency* and docker image bloat that would come with using puppeteer. as @PilsManuel mentioned cytosnap has been unmaintained for 5 years, whereas jsdom and node-canvas are mature and well-maintained.

even if cytosnap were to be maintained again, it's more code that needs to be actively maintained whereas it was fairly trivial to patch cytoscape to remove references to global window - the cy.window() method already existed and was simply not used in a few places in the render code path.

*I assume that the jsdom and node-canvas code paths are faster than running a headless browser, I haven't verified this myself but if needed I could create a benchmark comparing the performance of cytosnap versus my PR.

> There is currently some preliminary work being done on exporting graphs to PDF, which should also work in headless mode. Would that satisfy your requirements?

possibly? I presume rendering to a PDF would include some internal code to render to a raster or vector image to embed in said PDF, if this routine was accessible on its own through a public method, and the resulting image was stylistically identical to the canvas render, that might satisfy my requirements.

**Comment by mikekucera** [comment]
I've used cytosnap before and it was unreliable. I can understand why its not an option.

We've started on an extension for exporting the network to PDF using a vector image. But that project has just been started and probably won't be ready for a long while. 

Max is currently not available to manage the cytoscape.js project like he normally does. I'm managing the release process in the meantime. I apologize I'm not up to speed on the latest happenings in this repository. If you feel your PR (#3229) is tested and ready for being merged then we should be able to include it in the 3.29.1 release, which should be ready in a few days. Let me know what you think. Thanks.
